src/backend/distributed/planner/recursive_planning.c

```diff
@@ -1337,6 +1446,321 @@ NodeContainsSubqueryReferencingOuterQuery(Node *node)
+ * CreateOuterSubquery creates outer subquery which contains
+ * the given range table entry in its rtable.
+ */
+static Query *
+CreateOuterSubquery(RangeTblEntry *rangeTableEntry, List *allTargetList,
+					List *requiredAttrNumbers)
+{
+	MakeVarAttNosSequential(allTargetList);
+	List *innerSubqueryColNames = GenerateRequiredColNamesFromTargetList(allTargetList);
+
+	Query *outerSubquery = makeNode(Query);
+	outerSubquery->commandType = CMD_SELECT;
+
+	/* we copy the input rteRelation to preserve the rteIdentity */
+	RangeTblEntry *innerSubqueryRTE = copyObject(rangeTableEntry);
+
+	innerSubqueryRTE->eref->colnames = innerSubqueryColNames;
+	outerSubquery->rtable = list_make1(innerSubqueryRTE);
+
+	/* set the FROM expression to the subquery */
+	RangeTblRef *newRangeTableRef = makeNode(RangeTblRef);
+	newRangeTableRef->rtindex = 1;
+	outerSubquery->jointree = makeFromExpr(list_make1(newRangeTableRef), NULL);
+
+	outerSubquery->targetList = allTargetList;
+	return outerSubquery;
+}
+
+
+/*
+ * MakeVarAttNosSequential changes the attribute numbers of the given targetList
+ * to sequential numbers, [1, 2, 3] ...
+ */
+static void
+MakeVarAttNosSequential(List *targetList)
+{
+	TargetEntry *entry = NULL;
+	int attrNo = 1;
+	foreach_ptr(entry, targetList)
+	{
+		if (IsA(entry->expr, Var))
+		{
+			Var *var = (Var *) entry->expr;
+
+			/*
+			 * the inner subquery is an intermediate result hence
+			 * the attribute no's should be in ordinal order. [1, 2, 3...]
+			 */
+			var->varattno = attrNo++;
+		}
+	}
+}
+
+
+/*
+ * GenerateRequiredColNamesFromTargetList generates the required colnames
+ * from the given target list.
+ */
+static List *
+GenerateRequiredColNamesFromTargetList(List *targetList)
+{
+	TargetEntry *entry = NULL;
+	List *innerSubqueryColNames = NIL;
+	foreach_ptr(entry, targetList)
+	{
+		if (IsA(entry->expr, Var))
+		{
+			/*
+			 * column names of the inner subquery should only contain the
+			 * required columns, as in if we choose 'b' from ('a','b') colnames
+			 * should be 'a' not ('a','b')
+			 */
+			innerSubqueryColNames = lappend(innerSubqueryColNames, makeString(
+												entry->resname));
+		}
+	}
+	return innerSubqueryColNames;
+}
+
+
+/*
+ * UpdateVarNosInNode iterates the Vars in the
+ * given node and updates the varno's as the newVarNo.
+ */
+static void
+UpdateVarNosInNode(Query *query, Index newVarNo)
+{
+	List *varList = pull_var_clause(query->jointree->quals, PVC_RECURSE_AGGREGATES |
+									PVC_RECURSE_PLACEHOLDERS);
+	Var *var = NULL;
+	foreach_ptr(var, varList)
+	{
+		var->varno = newVarNo;
+	}
+}
+
+
+/*
+ * ContainsTableToBeConvertedToSubquery checks if the given range table list contains
+ * any table that should be converted to a subquery, which otherwise is not plannable.
+ */
+bool
+ContainsTableToBeConvertedToSubquery(List *rangeTableList)
+{
+	if (ContainsLocalTableDistributedTableJoin(rangeTableList))
+	{
+		return true;
+	}
+	if (ModifiesLocalTableWithRemoteCitusLocalTable(rangeTableList))
```

Comment:
Would putting remote Citus local tables into the "distributed"  candidates (perhaps renamed to remoteTableCandidates) be an option?

Reply:
I actually thought the same thing but didn't try, it seems like here we are checking is there a join between a local and a remote table. So I will try that